addons/bus/websocket.py

```diff
@@ -0,0 +1,883 @@
+        self.state = ConnectionState.CLOSING
+        initiated_by_client = not self._close_sent
+        if frame.code not in CLEAN_CLOSE_CODES or initiated_by_client:
+            self._terminate()
+        self._close_sent = True
+
+    def _process_next_message(self):
+        """
+        Process the next message coming throught the socket. If a
+        data message can be extracted, return its decoded payload.
+        As per the RFC, only control frames will be processed once
+        the connection reaches the closing state.
+        """
+        frame = self._get_next_frame()
+        if frame.opcode in CTRL_OP:
+            return self._handle_control_frame(frame)
+        if self.state is not ConnectionState.OPEN:
+            # After receiving a control frame indicating the connection
+            # should be closed, a peer discards any further data
+            # received.
+            return
+        if frame.opcode is Opcode.CONTINUE:
+            raise ProtocolError("Unexpected continuation frame")
+        message = frame.payload
+        if not frame.fin:
+            message = self._recover_fragmented_message(frame)
+        return (
+            message.decode('utf-8')
+            if message is not None and frame.opcode is Opcode.TEXT else message
+        )
+
+    def _recover_fragmented_message(self, initial_frame):
+        message_fragments = bytearray(initial_frame.payload)
+        while True:
+            frame = self._get_next_frame()
+            if frame.opcode in CTRL_OP:
+                # Control frames can be received in the middle of a
+                # fragmented message, process them as soon as possible.
+                self._handle_control_frame(frame)
+                if self.state is not ConnectionState.OPEN:
+                    return
+                continue
+            if frame.opcode is not Opcode.CONTINUE:
+                raise ProtocolError("A continuation frame was expected")
+            message_fragments.extend(frame.payload)
+            if len(message_fragments) > type(self).MESSAGE_MAX_SIZE:
+                raise PayloadTooLargeException()
+            if frame.fin:
+                return bytes(message_fragments)
+
+    def _send_frame(self, frame):
+        if frame.opcode in CTRL_OP and len(frame.payload) > 125:
+            raise ProtocolError(
+                "Control frames should have a payload length smaller than 126"
+            )
+        if isinstance(frame.payload, str):
+            frame.payload = frame.payload.encode('utf-8')
+        elif not isinstance(frame.payload, (bytes, bytearray)):
+            frame.payload = json.dumps(frame.payload).encode('utf-8')
+
+        output = bytearray()
+        first_byte = (
+              (0b10000000 if frame.fin else 0)
+            | (0b01000000 if frame.rsv1 else 0)
+            | (0b00100000 if frame.rsv2 else 0)
+            | (0b00010000 if frame.rsv3 else 0)
+            | frame.opcode
+        )
+        second_byte = 0
```

Comment:
Why bother with this? It's just being or-ed with actual content, so it never has any impact on the result.